Subman pipeline

kahowell opened the pull request:
Lots of TODOs; not ready for merge yet

Wanted to go ahead and push out what I had, so that folks can take a look and see if they like where this is going, etc.

@Lorquas @barnabycourt, take a look and let me know what you think. Feel free to tag others too.

Lorquas commented on jobs/submanPipeline.groovy:
```diff
@@ -0,0 +1,93 @@
+            body: '', // TODO fill in body
+            to: 'chainsaw@redhat.com',
+        )
+        // TODO figure out how to do IRC notification
```
should be similar to: https://github.com/candlepin/candlepin-jobs/blob/master/src/main/groovy/jobLib/rhsmLib.groovy#L42-L56

Lorquas commented:
Looks good so far to me, when we do the switch I guess we'll have to manually disable the old non-piplelined jobs?  Also is there a way to manually kick off the nodes/component tests of the pipeline individually?

kahowell commented:
> when we do the switch I guess we'll have to manually disable the old non-piplelined jobs

If we set the seed job to delete removed jobs, then it'll take care of itself. Or we could disable them with the DSL.

> is there a way to manually kick off the nodes/component tests of the pipeline individually?

hrmm... not as written... if this is a requirement, then we could keep the actual tests as separate jobs, and simply use the pipeline to orchestrate them. If you don't use `node`, then it doesn't tie up any executors. And you can run a job as a pipeline step. Actually, that might be a little easier to implement... I'll give it a shot and update this PR if it works out easily enough.

Ceiu commented:
Is it possible to use a job in a pipeline in multiple pipelines? If so, would it be possible to have another, parameterized pipeline which can invoke the test(s) we need?

kahowell commented on jobs/submanPipeline.groovy:
```diff
@@ -0,0 +1,79 @@
+                    body: "${subject}:\n\nCheck console output at ${env.BUILD_URL} to view the results",
+                    to: 'chainsaw@redhat.com',
+            )
+            // TODO figure out how to do IRC notification
```
A few options here:
1) this ugly shell script: https://jenkins.io/doc/pipeline/examples/#ircnotify-commandline
2) we could wrap the whole pipeline in the helper job (i.e. have the helper job block waiting on the pipeline to finish, and then have *it* do the IRC notify based on the pipeline result. Downside here is that it would occupy an extra executor.
3) Access the Jenkins IRC code from inside the pipeline script.
4) Hack on the IRC plugin?

Open to other ideas. @Lorquas @barnabycourt @Ceiu ?

barnabycourt replied:
@kahowell What about the native jenkins IRC plugin? https://wiki.jenkins-ci.org/display/JENKINS/IRC+Plugin

kahowell replied:
@barnabycourt there's probably a way to get at the functionality we want, but I haven't quite found it yet... Basically, I'm sure classes exist that do what we want, I'm just not sure they're easy to interface with (either way it'll definitely involve importing the classes into the pipeline scripts and invoking them directly, which is less than ideal).

kahowell commented:
> Is it possible to use a job in a pipeline in multiple pipelines?

Yes; my most recent changes might be what you're thinking of?

> If so, would it be possible to have another, parameterized pipeline which can invoke the test(s) we need?

Yes; we can even build in fancy logic for which to execute.

kahowell commented:
At a high level this does the following:

- Remove notifications from the jobs; instead we'll just notify from the
  pipeline.
- Remove triggers from the jobs; instead we'll have them triggered from
  the pipeline.
- Create a subscription-manager pipeline that uses the ghprb trigger,
  but does commit status notifications manually.

Could be improved:

- Hacky code to do IRC notifications; this makes the pipeline script
  require approval in Jenkins management.
- IRC notifications no longer trigger on fixed (Yippee... messages)
- Email does not notify on fixed.

Lorquas commented:
@kahowell This looks good to merge for me, should we go ahead or work on those 'could be improved' items?

kahowell commented:
@Lorquas, I'd like someone (besides myself) to at least verify that it all looks sane in vagrant first. I think we can work on all the improvements later though.

Lorquas commented:
@kahowell i rebased master into this branch, and put the work here: https://github.com/Lorquas/candlepin-jobs/tree/subman_pipeline

blocking me from testing the pipeline in vagrant, should i push those changes here and we can resolve that issue?

issue is:
Building in workspace /vagrant
Processing DSL script candlepinPerformanceTestsJob.groovy
ERROR: (candlepinPerformanceTestsJob.groovy, line 6) Custom config file with name 'candlepinPerformanceInventory' not found
Finished: FAILURE

kahowell committed:
Subman Pipeline

At a high level this does the following:

- Remove notifications from the jobs; instead we'll just notify from
the pipeline.
- Remove triggers from the jobs; instead we'll have them triggered
from the pipeline.
- Create a subscription-manager pipeline that uses the ghprb
trigger, but does commit status notifications manually.

Could be improved:

- Hacky code to do IRC notifications; this makes the pipeline script
  require approval in Jenkins management.
- IRC notifications no longer trigger on fixed (Yippee... messages)
- Email does not notify on fixed.

kahowell commented:
@Lorquas, I rebased on master, and resolved that issue. Notably, I added a variable: `jenkins_dummy_configs`; this is a list of names/ids (the script uses the `String` for both) of custom config files; I added the missing perf one there. You can vagrant destroy or you can vagrant provision to get it to run.

Lorquas committed:
adding to git ignore for ansible stuff

Lorquas commented:
@kahowell During vagrant provision I get
```
❯❯❯ vagrant provision
TASK [create dummy config files] ***********************************************
failed: [jenkins] (item=candlepinPerformanceInventory) => {"changed": true, "cmd": "echo \"javaposse.jobdsl.plugin.ConfigFileProviderHelper.findConfigProvider(javaposse.jobdsl.dsl.ConfigFileType.Custom).save(new org.jenkinsci.plugins.configfiles.custom.CustomConfig('candlepinPerformanceInventory', 'candlepinPerformanceInventory', 'dummy file', ''))\" | java -jar /opt/jenkins-cli.jar -s http://localhost:8080/ groovy = --username admin --password admin", "delta": "0:00:01.423781", "end": "2016-12-14 16:17:54.761718", "failed": true, "item": "candlepinPerformanceInventory", "rc": 1, "start": "2016-12-14 16:17:53.337937", "stderr": "\nERROR: Unexpected exception occurred while performing groovy command.\norg.codehaus.groovy.control.MultipleCompilationErrorsException: startup failed:\nRemoteClass: 1: unable to resolve class org.jenkinsci.plugins.configfiles.custom.CustomConfig \n @ line 1, column 118.\n   l.ConfigFileType.Custom).save(new org.je\n                                 ^\n\n1 error\n\n\tat org.codehaus.groovy.control.ErrorCollector.failIfErrors(ErrorCollector.java:310)\n\tat org.codehaus.groovy.control.CompilationUnit.applyToSourceUnits(CompilationUnit.java:946)\n\tat org.codehaus.groovy.control.CompilationUnit.doPhaseOperation(CompilationUnit.java:593)\n\tat org.codehaus.groovy.control.CompilationUnit.compile(CompilationUnit.java:542)\n\tat groovy.lang.GroovyClassLoader.doParseClass(GroovyClassLoader.java:298)\n\tat groovy.lang.GroovyClassLoader.parseClass(GroovyClassLoader.java:268)\n\tat groovy.lang.GroovyShell.parseClass(GroovyShell.java:688)\n\tat groovy.lang.GroovyShell.run(GroovyShell.java:517)\n\tat groovy.lang.GroovyShell.run(GroovyShell.java:497)\n\tat hudson.cli.GroovyCommand.run(GroovyCommand.java:86)\n\tat hudson.cli.CLICommand.main(CLICommand.java:264)\n\tat hudson.cli.CliManagerImpl.main(CliManagerImpl.java:93)\n\tat sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\n\tat sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)\n\tat sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\n\tat java.lang.reflect.Method.invoke(Method.java:606)\n\tat hudson.remoting.RemoteInvocationHandler$RPCRequest.perform(RemoteInvocationHandler.java:894)\n\tat hudson.remoting.RemoteInvocationHandler$RPCRequest.call(RemoteInvocationHandler.java:869)\n\tat hudson.remoting.RemoteInvocationHandler$RPCRequest.call(RemoteInvocationHandler.java:828)\n\tat hudson.remoting.UserRequest.perform(UserRequest.java:153)\n\tat hudson.remoting.UserRequest.perform(UserRequest.java:50)\n\tat hudson.remoting.Request$2.run(Request.java:336)\n\tat hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:68)\n\tat hudson.cli.CliManagerImpl$1.call(CliManagerImpl.java:64)\n\tat hudson.remoting.CallableDecoratorAdapter.call(CallableDecoratorAdapter.java:18)\n\tat hudson.remoting.CallableDecoratorList$1.call(CallableDecoratorList.java:21)\n\tat jenkins.util.ContextResettingExecutorService$2.call(ContextResettingExecutorService.java:46)\n\tat java.util.concurrent.FutureTask.run(FutureTask.java:262)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)\n\tat java.lang.Thread.run(Thread.java:745)", "stdout": "", "stdout_lines": [], "warnings": []}
	to retry, use: --limit @/home/jmolet/Projects/candlepin-jobs/ansible/vagrant.retry

PLAY RECAP *********************************************************************
jenkins                    : ok=29   changed=1    unreachable=0    failed=1   

Ansible failed to complete successfully. Any error output should be
visible above. Please fix these errors and try again.
```


which cleaned up looks like:

```
 echo "javaposse.jobdsl.plugin.ConfigFileProviderHelper.findConfigProvider(javaposse.jobdsl.dsl.ConfigFileType.Custom).save(new org.jenkinsci.plugins.configfiles.custom.CustomConfig('candlepinPerformanceInventory', 'candlepinPerformanceInventory', 'dummy file', ''))" | java -jar /opt/jenkins-cli.jar -s http://localhost:8080/ groovy = --username admin --password admin

ERROR: Unexpected exception occurred while performing groovy command.
org.codehaus.groovy.control.MultipleCompilationErrorsException: startup failed:
RemoteClass: 1: unable to resolve class org.jenkinsci.plugins.configfiles.custom.CustomConfig 
 @ line 1, column 118.
   l.ConfigFileType.Custom).save(new org.je
                                 ^

1 error

	at org.codehaus.groovy.control.ErrorCollector.failIfErrors(ErrorCollector.java:310)
	at org.codehaus.groovy.control.CompilationUnit.applyToSourceUnits(CompilationUnit.java:946)
	at org.codehaus.groovy.control.CompilationUnit.doPhaseOperation(CompilationUnit.java:593)
	at org.codehaus.groovy.control.CompilationUnit.compile(CompilationUnit.java:542)
	at groovy.lang.GroovyClassLoader.doParseClass(GroovyClassLoader.java:298)
	at groovy.lang.GroovyClassLoader.parseClass(GroovyClassLoader.java:268)
	at groovy.lang.GroovyShell.parseClass(GroovyShell.java:688)
	at groovy.lang.GroovyShell.run(GroovyShell.java:517)
	at groovy.lang.GroovyShell.run(GroovyShell.java:497)
	at hudson.cli.GroovyCommand.run(GroovyCommand.java:86)
	at hudson.cli.CLICommand.main(CLICommand.java:264)
	at hudson.cli.CliManagerImpl.main(CliManagerImpl.java:93)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at hudson.remoting.RemoteInvocationHandler$RPCRequest.perform(RemoteInvocationHandler.java:894)
	at hudson.remoting.RemoteInvocationHandler$RPCRequest.call(RemoteInvocationHandler.java:869)
	at hudson.remoting.RemoteInvocationHandler$RPCRequest.call(RemoteInvocationHandler.java:828)
	at hudson.remoting.UserRequest.perform(UserRequest.java:153)
	at hudson.remoting.UserRequest.perform(UserRequest.java:50)
	at hudson.remoting.Request$2.run(Request.java:336)
	at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:68)
	at hudson.cli.CliManagerImpl$1.call(CliManagerImpl.java:64)
	at hudson.remoting.CallableDecoratorAdapter.call(CallableDecoratorAdapter.java:18)
	at hudson.remoting.CallableDecoratorList$1.call(CallableDecoratorList.java:21)
	at jenkins.util.ContextResettingExecutorService$2.call(ContextResettingExecutorService.java:46)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
```

kahowell committed:
Add config-file-provider to vagrant

kahowell committed:
Install plugins before making config files

Lorquas committed:
throwing subman pipeline stuff into its own folder

Lorquas commented:
@kahowell I've added everything into a folder and it looks much cleaner, I can't test the run yet due to the script-security stuff, will mess with that later..  I'm good to merge this if you like my changes to this.
